### PR TITLE
Added HDFS functionality for streaming graphs from HDFS into Jasminegraph partitions.

### PR DESCRIPTION
**Pull request by muthumala19**
Added functionality to stream graph data from hdfs files and persist them in the Jasminegraph partitions.

### CODE REVIEW

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 1]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
     input_stream_handler_thread = thread(&StreamHandler::listen_to_kafka_topic, stream_handler);
 }
 
+void add_stream_hdfs_command(std::string masterIP,int connFd, std::string &hdfs_server_IP, std::thread &input_stream_handler_thread,
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1270]:
```suggestion
void addStreamHdfsCommand(std::string masterIP,int connFd, std::string &hdfsServerIP, std::thread &inputStreamHandlerThread,
```

Comment (miyurud) [reply src/frontend/JasmineGraphFrontEnd.cpp:1270]: There is issue I see with function names and variable names. Please follow the approach I have used in my comments.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 2]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+            return;
+        }
+
+        char file_path[FRONTEND_DATA_LENGTH + 1];
+        bzero(file_path, FRONTEND_DATA_LENGTH + 1);
+        read(connFd, file_path, FRONTEND_DATA_LENGTH);
+        std::string file_path_s(file_path);
+        file_path_s = Utils::trim_copy(file_path_s);
+
+        frontend_logger.info("Reading HDFS configuration file: " + file_path_s);
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1321]: Is this file part of JasmineGraph or is it something owned by the HDFS?

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 3]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+
+        std::vector<std::string> vec = Utils::getFileContent(file_path_s);
+        for (const auto &item : vec) {
+            frontend_logger.info("Item: " + item);
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1325]: I do not see a point of having log line here for the Item. Please remove it.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 4]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+            if (item.length() > 0 && !(item.rfind("#", 0) == 0)) {
+                std::vector<std::string> vec2 = Utils::split(item, '=');
+                if (vec2.size() == 2) {
+                    if (vec2.at(0).compare("hdfs.host") == 0) {
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1329]: This configuration file reading logic must be moved to some different class and different function. Do not clutter those details here.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 5]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+                    }
+                } else {
+                    frontend_logger.error("Invalid line in configuration file: " + item);
+                }
+            }
+        }
+    }
+
+    if (hdfs_server_IP.empty() || hdfs_port.empty()) {
+        frontend_logger.error("HDFS server IP or port not set correctly");
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1342]: This error message does not have any meaning. You need to point out the erroneous value and the reason for why you say there is an error.l

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 6]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+            if (item.length() > 0 && !(item.rfind("#", 0) == 0)) {
+                std::vector<std::string> vec2 = Utils::split(item, '=');
+                if (vec2.size() == 2) {
+                    if (vec2.at(0).compare("hdfs.host") == 0) {
+                        hdfs_server_IP = vec2.at(1);
+                    } else if (vec2.at(0).compare("hdfs.port") == 0) {
+                        hdfs_port = vec2.at(1);
+                    }
+                } else {
+                    frontend_logger.error("Invalid line in configuration file: " + item);
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1335]: Which configuration file are you referring to here?

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 7]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+
+    if (hdfs_server_IP.empty() || hdfs_port.empty()) {
+        frontend_logger.error("HDFS server IP or port not set correctly");
+        *loop_exit_p = true;
+        return;
+    }
+
+    frontend_logger.info("HDFS server IP: " + hdfs_server_IP + " HDFS port: " + hdfs_port);
+
+    frontend_logger.info("Start serving HDFS command");
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1349]: No meaning in this info log

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 8]
```diff
@@ -1251,6 +1267,137 @@ static void add_stream_kafka_command(int connFd, std::string &kafka_server_IP, c
...
+    std::string message = "Send HDFS file path";
```
Comment (miyurud) [inline src/frontend/JasmineGraphFrontEnd.cpp:1350]: You should have short word commands than having this kind of lengthy sentences as the commands.

File: src/partitioner/local/MetisPartitioner.cpp [hunk 9]
```diff
@@ -1009,3 +1009,125 @@ void MetisPartitioner::writeSerializedCompositeMasterFiles(std::string part) {
     masterFileMutex.unlock();
     partitioner_logger.log("Serializing done for central part " + part, "info");
 }
+
+void MetisPartitioner::setPathAndGraphID(int graphID) {
+    this->outputFilePath = Utils::getHomeDir() + "/.jasminegraph/tmp/" + std::to_string(graphID);
+    this->graphID = graphID;
+}
+
+void MetisPartitioner::loadFileChunks(string inputFilePath) {
```
Comment (miyurud) [inline src/partitioner/local/MetisPartitioner.cpp:1018]: This function name is misleading. It seems it does not do any meaningful processing.

File: src/partitioner/local/MetisPartitioner.cpp [hunk 10]
```diff
@@ -1009,3 +1009,125 @@ void MetisPartitioner::writeSerializedCompositeMasterFiles(std::string part) {
     masterFileMutex.unlock();
     partitioner_logger.log("Serializing done for central part " + part, "info");
 }
+
+void MetisPartitioner::setPathAndGraphID(int graphID) {
```
Comment (miyurud) [inline src/partitioner/local/MetisPartitioner.cpp:1013]: This function is not useful. I do not see any point of having such function.

File: src/frontend/JasmineGraphFrontEnd.cpp [hunk 11]
```diff
@@ -136,6 +141,10 @@ void *frontendservicesesion(void *dummyPt) {
     KafkaConnector *kstream;
     Partitioner graphPartitioner(numberOfPartitions, 1, spt::Algorithms::HASH);
 
+    //Initiate HDFS parameters
+    std::string hdfs_server_IP;
```
[the remark this answers is not in the document]
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Changed variable names, and method names into camel case notation.

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 1 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Changed variable names, and method names into camel case notation.

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 3 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Removed.

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 5 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Added more specific error handling with below code,

    if (hdfsServerIp.empty()) {
        frontend_logger.error("HDFS server IP is not set or empty.");
    }
    if (hdfsPort.empty()) {
        frontend_logger.error("HDFS server port is not set or empty.");
    }

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 6 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp:1330]: In this hdfs implementation, there are two types of HDFS servers.
1. Jasminegraph's Default HDFS server.
2. Custom HDFS servers from users.
users can use either one of those two types. if a user wants to refer to a custom hdfs server, then he needs to provide config file to that HDFS server. Here it refers to the custom hdfs server config file.

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 7 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Removed.

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 8 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: Shortened as "HDFS file path: "

File: src/partitioner/local/MetisPartitioner.cpp (hunk 9 above)
Comment (muthumala19) [reply src/partitioner/local/MetisPartitioner.cpp:1018]: This method is no longer there in the updated hdfs implementation.

File: src/partitioner/local/MetisPartitioner.cpp (hunk 10 above)
Comment (muthumala19) [reply src/partitioner/local/MetisPartitioner.cpp:1013]: This method is no longer there in the updated hdfs implementation.

File: src/frontend/JasmineGraphFrontEnd.cpp (hunk 2 above)
Comment (muthumala19) [reply src/frontend/JasmineGraphFrontEnd.cpp]: In this hdfs implementation, there are two types of HDFS servers.
1. Jasminegraph's Default HDFS server.
2. Custom HDFS servers from users.
users can use either one of those two types. if a user wants to refer to a custom hdfs server, then he needs to provide config file to that HDFS server. Here it refers to the custom hdfs server config file. currently this file should be put into the jasminegraph docker container and give the respective path.

**Comment by muthumala19** [comment]
closing this PR due to open another PR. (#268)

**Pull request closed without merging**